KAFKA-13173 Making the controller fence stale brokers one at a time if multiple stale brokers are discovered at the same time

niket-goel opened the pull request:
This is necessary as the effect of each fencing is visible to the
controller one the corresponding record for it has been generated and
applied. Processing stale brokers one at a time ensures that subsequent 
fencing records are generated correctly with the correct view of the system.

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -858,7 +858,7 @@ private void rescheduleMaybeFenceStaleBrokers() {
         scheduleDeferredWriteEvent(MAYBE_FENCE_REPLICAS, nextCheckTimeNs, () -> {
-            ControllerResult<Void> result = replicationControl.maybeFenceStaleBrokers();
+            ControllerResult<Void> result = replicationControl.maybeFenceOneStaleBroker();
             rescheduleMaybeFenceStaleBrokers();
```
One thing I was considering is whether we should schedule the next check right away. For example, maybe we could check whether `result` contains a non-empty record set and reschedule immediately (or perhaps after a short backoff) if it does.

niket-goel replied:
I thought about doing that, but then realized that a call to reschedule will automatically result in a scheduling of another invocation right away if there were any brokers that were deemed stale still remaining in the unfenced list.

This will happen because the `nextCheckTimeNs` is based on when the broker with the older `lastContactNs` time (broker at the head of the list) would time out. This will always result in a time less than now if we have a stale broker and there should be no delay in executing this again (barring queue depth).

I also did not want to "prioritize" (if that's even possible) over other requests, because my understanding is that the fix works as long as a subsequent fence broker call is made after the first has been processed by the controller thread (which would also happen by picking up the record from the queue and replaying it).

Please let me know if my understanding is off here.

hachikuji replied:
It sounds like you are saying there may still be a race condition if the next expiration is scheduled too quickly. However, I think the way the code works, we are guaranteed to have this result reflected before the next scheduled operation can trigger. It would be good to have a test covering this scenario.

niket-goel replied:
I agree that a test covering this would be good. Will add that to the next revision.

niket-goel replied:
added tests in the next revision.

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -925,14 +926,21 @@ ApiError electLeader(String topic, int partitionId, boolean uncleanOk,
         BrokerHeartbeatManager heartbeatManager = clusterControl.heartbeatManager();
         List<Integer> staleBrokers = heartbeatManager.findStaleBrokers();
-        for (int brokerId : staleBrokers) {
+        if (!staleBrokers.isEmpty()) {
```
If we are only expiring one, maybe we can change `findStaleBrokers` to `findStaleBroker` and return `Optional<Integer>` instead of `List<Integer>`.

niket-goel replied:
That is fair. Let me do that (wasn't sure if we wanted to add another method for this case).

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -925,14 +926,21 @@ ApiError electLeader(String topic, int partitionId, boolean uncleanOk,
             log.info("Fencing broker {} because its session has timed out.", brokerId);
             handleBrokerFenced(brokerId, records);
             heartbeatManager.fence(brokerId);
+
```
nit: remove newline

niket-goel replied:
ack

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -419,6 +419,27 @@ long nextCheckTimeNs() {
+            // The unfenced broker list is sorted on last contact time from each
+            // broker. If the first broker has a valid session then all do
+            if (!hasValidSession(broker)) {
+                node = Optional.of(broker.id);
```
nit: you can return here and then we don't need `node`

niket-goel replied:
will do.

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -925,18 +925,27 @@ ApiError electLeader(String topic, int partitionId, boolean uncleanOk,
-        List<Integer> staleBrokers = heartbeatManager.findStaleBrokers();
-        for (int brokerId : staleBrokers) {
+        Optional<Integer> staleBroker = heartbeatManager.findOneStaleBroker();
+        if (staleBroker.isPresent()) {
```
nit: maybe a little more concise with a lambda?
```java
heartbeatManager.findOneStaleBroker().ifPresent(brokerId -> {
  // Even though multiple brokers can go stale at a time, we will process
  // fencing one at a time so that the effect of fencing each broker is visible
  // to the system prior to processing the next one
  log.info("Fencing broker {} because its session has timed out.", brokerId);
  handleBrokerFenced(brokerId, records);
  heartbeatManager.fence(brokerId);
});
```

niket-goel replied:
Will do.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/BrokerHeartbeatManagerTest.java:
```diff
@@ -93,22 +93,26 @@ public void testFindStaleBrokers() {
-        manager.fence(2);
-        assertEquals(Collections.emptyList(), manager.findStaleBrokers());
+        Integer nodeId = 1;
+        while (manager.findOneStaleBroker().isPresent()) {
```
nit: maybe it's better to unroll this to make the expectation explicit. This test would pass even if there were additional iterations of the loop. 
```java
assertEquals(Optional.of(1), manager.findStaleBroker());
manager.fence(1);

assertEquals(Optional.of(2), manager.findStaleBroker());
manager.fence(2);
```
Ditto below.

niket-goel replied:
ack. Good catch! I should have added an assert on the node count.  
Will just unroll the loop .

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTestEnv.java:
```diff
@@ -37,7 +42,13 @@
 
     public QuorumControllerTestEnv(LocalLogManagerTestEnv logEnv,
-                                   Consumer<QuorumController.Builder> builderConsumer)
+        Consumer<QuorumController.Builder> builderConsumer)
```
nit: the indentation is a bit screwy here and below

niket-goel replied:
ack. Will match the indentation to what it was before.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/ReplicationControlManagerTest.java:
```diff
@@ -1029,6 +1051,22 @@ public void testCreatePartitionsFailsWithManualAssignmentWithAllFenced() throws
+        ctx.fenceBrokers(Utils.mkSet(2, 3));
+
+        PartitionRegistration partition0 = replication.getPartition(fooId, 0);
+        assertArrayEquals(new int[]{1}, partition0.isr);
```
Can we extend this test to assert something about the other partitions? Maybe also useful to assert the current leader?

niket-goel replied:
Makes sense. Will add the leader assertion and see what other things we can assert here.

niket-goel committed:
KAFKA-13173 Making the controller fence stale brokers one at a time if
multiple stale brokers are discovered at the same time

This is necessary as the effect of each fencing is visible to the
controller one the corresponding record for it has been generated and
applied. Processing stale brokers one at a time ensures that subsequent
fencing records are generated correctly with the correct view of the system.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,112 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
     }
 
+    @Test
+    public void testFenceMultipleBrokers() throws Throwable {
```
This test times out for me when run locally. I'm a little concerned that the dependence on real time here will make it flaky. I'm satisfied with the other tests we have in this patch. Would it be reasonable to push this test case to a follow-up so that we can iterate on it a little bit without blocking the patch?

niket-goel replied:
The timeout was because I accidentally deleted a line before pushing the commit. Will fix that. 
Will discuss pulling the test out offline with you.

niket-goel committed:
Addressing PR comments

niket-goel committed:
Fixing style violations caught by Jenkins

showuon commented on metadata/src/test/java/org/apache/kafka/controller/ReplicationControlManagerTest.java:
```diff
@@ -1029,6 +1051,24 @@ public void testCreatePartitionsFailsWithManualAssignmentWithAllFenced() throws
+
+        ctx.fenceBrokers(Utils.mkSet(2, 3));
+
+        PartitionRegistration partition0 = replication.getPartition(fooId, 0);
+        assertArrayEquals(new int[]{1, 2, 3}, partition0.replicas);
+        assertArrayEquals(new int[]{1}, partition0.isr);
+        assertEquals(1, partition0.leader);
```
I think we can also verify other partitions, not just partition 0. ex:
```java
        PartitionRegistration partition0 = replication.getPartition(fooId, 0);
        PartitionRegistration partition1 = replication.getPartition(fooId, 1);
        PartitionRegistration partition2 = replication.getPartition(fooId, 2);
        assertArrayEquals(new int[]{1, 2, 3}, partition0.replicas);
        assertArrayEquals(new int[]{1}, partition0.isr);
        assertEquals(1, partition0.leader);
        assertArrayEquals(new int[]{2, 3, 4}, partition1.replicas);
        assertArrayEquals(new int[]{4}, partition1.isr);
        assertEquals(4, partition1.leader);
        assertArrayEquals(new int[]{0, 2, 1}, partition2.replicas);
        assertArrayEquals(new int[]{0, 1}, partition2.isr);
        assertNotEquals(2, partition2.leader);
```

showuon commented on metadata/src/test/java/org/apache/kafka/controller/ReplicationControlManagerTest.java:
```diff
@@ -208,6 +212,24 @@ void unfenceBrokers(Integer... brokerIds) throws Exception {
         }
 
+        void fenceBrokers(Set<Integer> brokerIds) throws Exception {
+            time.sleep(BROKER_SESSION_TIMEOUT_MS);
```
nit: we can assert the `fencedBrokerIds` is empty before sleep to timeout session.

niket-goel replied:
Good call.. Although I would not do it in the common utility function. 
Will do it in the test method where this should hold.

showuon commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +175,127 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+                    waitIterations++;
+
+                    if (waitIterations >= sessionTimeout * 3) {
+                        assertTrue(false, "Fencing of brokers did not process within expected time");
```
You can call `fail` directly: `fail("Fencing of brokers did not process within expected time");`

niket-goel replied:
Thanks (new to the project and JAVA :) )

showuon commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +175,127 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+                    if (waitIterations >= sessionTimeout * 3) {
+                        assertTrue(false, "Fencing of brokers did not process within expected time");
+                    }
+                } while (!fencingComplete);
```
Maybe you can replace this `while` loop with a util function `TestUtils.waitForCondition`. Ex:
`TestUtils.waitForCondition(() -> fenceBrokers(), 3000, "Fencing of brokers did not process within expected time");`
Just that it only accepted the `timeout` value, not the `waitIterations` like you did. FYI

niket-goel replied:
Will check this out. Thanks for the pointer.

showuon reviewed (commented):
@niket-goel , thanks for the PR. Left some comments for tests. Thanks.

niket-goel commented:
@showuon -- Thanks for looking at the PR. Will address the comments and publish an update.

niket-goel committed:
Modifying tests based on some PR feedback

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -420,22 +419,22 @@ long nextCheckTimeNs() {
+    Optional<Integer> findOneStaleBroker() {
         BrokerHeartbeatStateIterator iterator = unfenced.iterator();
-        while (iterator.hasNext()) {
+        if (iterator.hasNext()) {
```
Isn't `unfenced.first()` suppose to return the oldest heartbeat? If so, can use that method and check for `null`?

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -161,6 +163,14 @@ public void deactivate() {
         return brokerRegistrations;
     }
 
+    Set<Integer> fencedBrokerIds() {
```
Looks like this method is only used for tests. How about moving this to `ReplicationControlManagerTest`?

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,121 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+                                setReplicationFactor(replicationFactor)).iterator()));
+                CreateTopicsResponseData createTopicsResponseData = active.createTopics(
+                    createTopicsRequestData).get();
+                assertEquals(Errors.INVALID_REPLICATION_FACTOR.code(),
```
This part is a little mysterious to me. We return an error, but the topic is created anyway? That seems surprising.

niket-goel replied:
The topic creation actually fails here. We try the create again after unfencing the brokers.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,121 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+                sendBrokerheartbeat(active, brokerCount, brokerEpochs);
+                createTopicsResponseData = active.createTopics(
+                    createTopicsRequestData).get();
+                assertEquals(Errors.NONE.code(), createTopicsResponseData.topics().find("foo").errorCode());
```
nit: it's a small thing, but the assertion failure message is more useful if we use the `Errors` type.
```java
assertEquals(Errors.NONE, Errors.forCode(createTopicsResponseData.topics().find("foo").errorCode()));
```

niket-goel replied:
Makes sense, will modify.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,121 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+    @Test
+    public void testFenceMultipleBrokers() throws Throwable {
+        int brokerCount = 5;
+        int brokersToKeepUnfenced = 1;
```
Maybe a little subjective, but I think the test case would be more readable if we list these brokers directly. For example:
```java
List<Integer> allBrokers = Arrays.asList(1, 2, 3, 4, 5);
List<Integer> brokersToKeepUnfenced = Arrays.asList(1);
```

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,121 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+                        new CreatableTopicCollection(Collections.singleton(
+                            new CreatableTopic().setName("bar").setNumPartitions(1).
+                                setConfigs(new CreateableTopicConfigCollection(Collections.
+                                    singleton(new CreateableTopicConfig().setName("min.insync.replicas").
```
Is there a point to setting `min.insync.replicas` in this test? I am wondering why we don't just reuse the original create request.

niket-goel replied:
TL;DR Is vestigial, will remove.

The reason i added it in there was because even with only one broker available the topic creation would succeed. This added condition of a min.isr was to strengthen the topic creation criteria to make it fail when enough brokers were not un-fenced. I learnt that today the behavior is to not block topic creation even if the min.isr would not be met for the topic on a subsequent produce. Anyway, unless that behavior is changed the min.isr is serving no purpose and I will remove it. I added explicit checks to ensure broker liveness.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,121 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+        int brokerCount = 5;
+        int brokersToKeepUnfenced = 1;
+        short replicationFactor = 5;
+        Long sessionTimeoutSec = 1L;
```
nit: maybe we can use milliseconds consistently and leave out some unit conversions below.

Also, could this and `sleepMillis` be a `long` instead of `Long`?

niket-goel replied:
Will make the change. Had read the session timeout parameter wrong. Will keep consistent time units.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,121 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+        Long sleepMillis = (sessionTimeoutSec * 1000) / 2;
+
+        try (LocalLogManagerTestEnv logEnv = new LocalLogManagerTestEnv(1, Optional.empty())) {
+            try (QuorumControllerTestEnv controlEnv =
```
nit: we can combine these two `try` blocks:
```java
try (LocalLogManagerTestEnv logEnv = new LocalLogManagerTestEnv(1, Optional.empty());
     QuorumControllerTestEnv controlEnv = new QuorumControllerTestEnv(logEnv, b -> b.setConfigDefs(CONFIGS), Optional.of(sessionTimeoutSec))) {
...
```

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,121 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+                        new BrokerRegistrationRequestData().
+                            setBrokerId(brokerId).
+                            setClusterId("06B-K3N1TBCNYFgruEVP0Q").
+                            setIncarnationId(Uuid.fromString("kxAT73dKQsitIedpiPtwBA")).
```
Perhaps we can use `Uuid.randomUuid`? It's a little weird for all brokers to have the same incarnationId.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +176,121 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+                // At this point the brokers we want fenced are fenced.
+                // Send another heartbeat to the brokers we want to keep alive
+                sendBrokerheartbeat(active, brokersToKeepUnfenced, brokerEpochs);
+                int[] expectedIsr = new int[brokersToKeepUnfenced];
```
If we create the topic with an explicit assignment, then I think we can remove this logic to build the expected ISR since we would be able to assert the expected ISR directly.

niket-goel replied:
Given we are keeping only one broker to remain unfenced, the expected isr and leader are simply that broker. I kept this logic to allow for paramterization of the test. Given we are not using the parameters right now, will remove this logic.

hachikuji commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTestEnv.java:
```diff
@@ -36,16 +41,29 @@
+    public QuorumControllerTestEnv(
+        LocalLogManagerTestEnv logEnv,
+        Consumer<Builder> builderConsumer,
+        Optional<Long> sessionTimeoutSeconds
```
nit: as above, can we use milliseconds? It's the most conventional unit we use in the code for configurations representing durations.

niket-goel replied:
Will do.

niket-goel committed:
Addressing PR comments around TestFenceMultipleBrokers

niket-goel committed:
More simplifications to the test case

niket-goel committed:
Fixing comment typo

hachikuji reviewed (approved):
LGTM. Thanks for the fix!

showuon reviewed (approved):
LGTM! Thanks for the update.

jsancio commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -173,6 +174,89 @@ private void testDelayedConfigurationOperations(LocalLogManagerTestEnv logEnv,
+
+            assertTrue(Arrays.equals(isrFoo, expectedIsr),
+                "The ISR for topic foo was " + Arrays.toString(isrFoo) +
+                    ". It is expected to be " + Arrays.toString(expectedIsr));
```
If you use `assertEquals` it will compare the content of the arrays and print their value if they don't match.

jsancio commented on metadata/src/test/java/org/apache/kafka/controller/ReplicationControlManagerTest.java:
```diff
@@ -1029,6 +1052,36 @@ public void testCreatePartitionsFailsWithManualAssignmentWithAllFenced() throws
+        Uuid fooId = ctx.createTestTopic("foo", new int[][]{
+            new int[]{1, 2, 3}, new int[]{2, 3, 4}, new int[]{0, 2, 1}}).topicId();
+
+        assertTrue(ctx.clusterControl.fencedBrokerIds().isEmpty());
```
You can `assertEquals` against the empty `Set` (`Collections.emptySet()`). When it fails it will print the content of fenced broker ids.

jsancio reviewed (commented):
Thanks for the changes they look good in general. Apologies for the late review but I had some minor suggestions.

hachikuji commented:
@niket-goel I'm inclined to merge this as is so that we don't need to wait for another build. Perhaps you can submit a small follow-up to address @jsancio 's comments?

niket-goel commented:
@hachikuji Will follow up with a follow-up. Thanks for merging this.
